Makefile

```diff
@@ -1,277 +1,251 @@
-	ln -sf ${IDRIS2_CURDIR}/libs/base/build/ttc    ${TEST_PREFIX}/${NAME_VERSION}/base-${IDRIS2_VERSION}
-	ln -sf ${IDRIS2_CURDIR}/libs/test/build/ttc    ${TEST_PREFIX}/${NAME_VERSION}/test-${IDRIS2_VERSION}
-	ln -sf ${IDRIS2_CURDIR}/libs/contrib/build/ttc ${TEST_PREFIX}/${NAME_VERSION}/contrib-${IDRIS2_VERSION}
-	ln -sf ${IDRIS2_CURDIR}/libs/network/build/ttc ${TEST_PREFIX}/${NAME_VERSION}/network-${IDRIS2_VERSION}
+  $(info Idris Makefile invoked; version [$(IDRIS_VERSION)], version tag [$(IDRIS_VERSION_TAG)].)
 endif
 
-.PHONY: ${TEST_PREFIX}/${NAME_VERSION}
-
-testenv:
-	@${MAKE} ${TEST_PREFIX}/${NAME_VERSION}
-	@${MAKE} -C tests testbin IDRIS2=${TARGET} IDRIS2_PREFIX=${TEST_PREFIX}
-
-testenv-clean:
-	$(RM) -r ${TEST_PREFIX}/${NAME_VERSION}
-
-test: testenv
-	@echo
-	@echo "NOTE: \`${MAKE} test\` does not rebuild Idris or the libraries packaged with it; to do that run \`${MAKE}\`"
-	@if [ ! -x "${TARGET}" ]; then echo "ERROR: Missing IDRIS2 executable. Cannot run tests!\n"; exit 1; fi
-	@echo
-	@${MAKE} -C tests only=$(only) IDRIS2=${TARGET} IDRIS2_PREFIX=${TEST_PREFIX}
-
-retest: testenv
-	@echo
-	@echo "NOTE: \`${MAKE} retest\` does not rebuild Idris or the libraries packaged with it; to do that run \`${MAKE}\`"
-	@if [ ! -x "${TARGET}" ]; then echo "ERROR: Missing IDRIS2 executable. Cannot run tests!\n"; exit 1; fi
-	@echo
-	@${MAKE} -C tests retest only=$(only) IDRIS2=${TARGET} IDRIS2_PREFIX=${TEST_PREFIX}
-
-test-installed:
-	@${MAKE} -C tests testbin      IDRIS2=$(IDRIS2_PREFIX)/bin/idris2 IDRIS2_PREFIX=${IDRIS2_PREFIX}
-	@${MAKE} -C tests only=$(only) IDRIS2=$(IDRIS2_PREFIX)/bin/idris2 IDRIS2_PREFIX=${IDRIS2_PREFIX}
-
-support:
-	@${MAKE} -C support/c
-	@${MAKE} -C support/refc
-	@${MAKE} -C support/chez
+export IDRIS_FULL_VERSION := $(IDRIS_VERSION)$(if $(IDRIS_VERSION_TAG),-$(IDRIS_VERSION_TAG))
+
+define make-stage-generic
+  $(MAKE) --makefile Makefile-one-stage.mk \
+    BOOTSTRAP_IDRIS="$(if $(1),$(abspath $(1)/bin/$(NAME)),$(BOOTSTRAP_IDRIS))" \
+    BUILD=$(2) \
+    PREFIX="$(3)" \
+    IDRIS2_CG=$(IDRIS2_CG) \
+    $(4) $(5) $(6) $(7) $(8)
+endef
+
+define make-stage
+  $(call make-stage-generic,$(if $(1),$(1)/installed),$(2),$(abspath $(2)/installed),$(3),$(4),$(5),$(6),$(7),$(8))
+endef
+
+all: stage2
+
+$(BUILD):
+	mkdir -p $@
+
+$(STAGE0)/src-$(STAGE0_GIT_REF):
+	git worktree add --detach --force $@ $(STAGE0_GIT_REF)
+
+stage0 $(STAGE0_IDRIS): $(STAGE0)/src-$(STAGE0_GIT_REF)
+	@echo -e "\n*** Building the bootstrap host (aka stage 0); STAGE0_GIT_REF=$(STAGE0_GIT_REF)\n"
+	$(MAKE) --directory $< SCHEME="$(CHEZ)" PREFIX="$(abspath $(STAGE0)/installed)" bootstrap install
+
+# We don't FORCE stage1, which means that it only gets rebuilt after a
+# make clean. Also note that stage0 is only part of the prerequisits
+# when we are not taking a bootstrap shortcut.
+stage1 $(STAGE1_IDRIS): $(STAGE1_PREREQUISITES)
+	@echo -e "\n*** Building stage 1 using [$(or $(STAGE0_GIT_REF),\
+$(BOOTSTRAP_IDRIS),bootstrap files)]\n"
```

Comment:
no. it's using the `$(or ...)` mechanism of the initial, string-substitution phase of the makefile. if we put whitespaces there then it will show up in the output of that `echo`. AFAIU the only other option is to keep it as one line.